pkg/services/ngalert/schedule/schedule.go

```diff
@@ -487,10 +509,9 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
 				}
 			}()
 		case <-grafanaCtx.Done():
-			// clean up the state only if the reason for stopping the evaluation loop is that the rule was deleted
-			if errors.Is(grafanaCtx.Err(), errRuleDeleted) {
-				clearState()
-			}
+			// We need to use context.Background() here because grafanaCtx is being canceled, and we need a valid context
+			// in clearState function to do DB operations.
+			clearState(context.Background(), grafanaCtx.Err())
```

Comment:
If `context.Background()` is being used in the database queries then we could have an infinite timeout? @yuri-tceretian

Reply:
I just had another look at the code and I think we should add a timeout here. Perhaps something like:

```suggestion
			ctx, cancelFunc := context.WithTimeout(context.Background(), 5 * time.Second)
			defer cancelFunc()
			clearState(ctx, grafanaCtx.Err())
```